File: Life/Tests/DomainTests/Perfect/DomainTest.cs
#region Usings

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Domain;
using Microsoft.VisualStudio.TestTools.UnitTesting;

#endregion

namespace DomainTests.Perfect {
    public class DomainTest {
        public DomainTest() {
            Create = new Father();
        }

        public class AssertThat {
            public static void AreEqual(string expectedGameRepresentation, Game game) {
                var row = 0;
                var lines = expectedGameRepresentation.Split(new[] {Environment.NewLine, "\n"}, StringSplitOptions.RemoveEmptyEntries).ToList().ConvertAll(_ => _.Trim());
                var errors = expectedGameRepresentation.Split(new[] {Environment.NewLine, "\n"}, StringSplitOptions.RemoveEmptyEntries).ToList().ConvertAll(_ => _.Trim().ToCharArray());
                var hasErrors = false;

                foreach (var line in lines) {
                    var column = 0;
                    foreach (var code in line) {
                        switch (code) {
                            case '.':
                                if (game[row, column].IsAlive) {
                                    hasErrors = true;
                                    errors[row][column] = '*';
                                }
                                break;
                            case 'x':
                            case 'X':
                                if (game[row, column].IsDead) {
                                    hasErrors = true;
                                    errors[row][column] = '*';
                                }
                                break;
                        }
                        column++;
                    }
                    row++;
                }
                if (hasErrors) {
                    Assert.Fail("Games doesn't match:" + Environment.NewLine + ConvertToString(errors));
                }
            }

            private static string ConvertToString(IEnumerable<char[]> errors) {
                var sb = new StringBuilder();
                foreach (var error in errors) {
                    sb.AppendLine(new string(error));
                }
                return sb.ToString();
            }
        }

        public Father Create { get; private set; }
    }
}

File: life-java/src/test/java/ru/agiledays/life/server/domain/DomainTest.java
package ru.agiledays.life.server.domain;

import org.junit.Assert;
import static ru.agiledays.life.server.domain.StringHelper.*;

public class DomainTest {
    public DomainTest() {
        create = new Father();
    }
    
    protected static class AssertThat {
        public static void areEqual(final String expectedGameRepresentation, final Game game) {
            int row = 0;
            final String[] lines = splitLines(expectedGameRepresentation);
            final char[][] errors = asCharMap(lines);
            boolean hasErrors = false;

            for (String line: lines) {
                int column = 0;
                for (char code: line.toCharArray()) {
                    switch (code) {
                        case '.':
                            if (game.getCell(row, column).isAlive()) {
                                hasErrors = true;
                                errors[row][column] = '*';
                            }
                            break;
                        case 'x':
                        case 'X':
                            if (game.getCell(row, column).isDead()) {
                                hasErrors = true;
                                errors[row][column] = '*';
                            }
                            break;
                    }
                    column++;
                }
                row++;
            }
            if (hasErrors) {
                Assert.fail("Games doesn't match:" + newLine + convertToString(errors));
            }
        }
        
        private static String convertToString(final char[][] errors) {
            final StringBuffer sb = new StringBuffer();
            for (char[] error: errors) {
                sb.append(new String(error));
                sb.append(newLine);
            }
            return sb.toString();
        }
    }
    
    protected final Father create;
}
